bionic/cache.py

```diff
@@ -746,16 +747,15 @@ def _validated_object_name_from_url(self, url):
         return object_name
 
     def _bucket_and_object_names_from_url(self, url):
-        if not url.startswith(self._GS_URL_PREFIX):
-            raise ValueError(f'url must start with "{self._GS_URL_PREFIX}"')
-        url_parts = url[len(self._GS_URL_PREFIX):].split('/', 1)
-        if len(url_parts) == 1:
-            bucket_name, = url_parts
-            object_prefix = ''
-        else:
-            bucket_name, object_prefix = url_parts
-
-        return bucket_name, object_prefix
+        if not is_gcs_url(url):
+            raise ValueError(f'url must have schema "{GCS_SCHEME}"')
+        result = urlparse(url)
+        object_name = result.path
+        # urlparse parses the url with a leading slash in path
+        # we don't want to write to the root dir in gcs bucket
+        if object_name.startswith('/'):
```

Comment:
I'm confused by this part. `urlparse` only includes a leading slash if there's an extra slash in the URL, right? Like `gs:///bucket/dir/data.txt`? But isn't this actually a different (and maybe invalid) URL compared to `gs://bucket/dir/data.txt`?

Also, when I run `urlparse('gs:///bucket/dir/data.txt')`, it leaves `netloc` empty, so that seems like it'd be a problem too.

Reply:
I don't think that's true. IIUC, a url like `scheme://netloc/path` are absolute urls so the path can never be relative. I think that's why the url always contains a leading slash. But for gcs, it seems like Bionic assumes that the path is always relative to the cache dir. Hence I'm removing the leading slash. Maybe I can add this to the comment for clarity.

If you see the [syntax spec](https://cloud.google.com/storage/docs/gsutil#syntax), it is `gs://[BUCKET_NAME]/[OBJECT_NAME]`. So I guess having 3 slashes means there is no bucket. Is that a valid url or something you have seen before?

Reply:
Sorry, as we discussed, I was confused here. I think the behavior makes sense, although the comment may be misleading.

Oh, but: if `urlparse` *always* returns a leading slash, maybe it would be better to have an `assert` rather than an `if`?